shaders/Utility/ColorArrayInputAdapter.osl

```diff
@@ -0,0 +1,108 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2023, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferOSL/ArrayArgList.h"
+
+shader ColorArrayInputAdapter
+(
+	color in0 = 0,
+	color in1 = 0,
+	color in2 = 0,
+	color in3 = 0,
+	color in4 = 0,
+	color in5 = 0,
+	color in6 = 0,
+	color in7 = 0,
+	color in8 = 0,
+	color in9 = 0,
+	color in10 = 0,
+	color in11 = 0,
+	color in12 = 0,
+	color in13 = 0,
+	color in14 = 0,
+	color in15 = 0,
+	color in16 = 0,
+	color in17 = 0,
+	color in18 = 0,
+	color in19 = 0,
+	color in20 = 0,
+	color in21 = 0,
+	color in22 = 0,
+	color in23 = 0,
+	color in24 = 0,
+	color in25 = 0,
+	color in26 = 0,
+	color in27 = 0,
+	color in28 = 0,
+	color in29 = 0,
+	color in30 = 0,
+	color in31 = 0,
+
+	output color out1[1] = { INLIST_1 },
```

Comment:
Have you done any testing to check that the OSL optimiser does a good job of ignoring all the unused variants?

Reply:
Looks like it's fine, at least after optimization:

Running 
`env OSL_OPTIONS="debug=1" gaffer test GafferOSLTest`

yields:

```After optimizing layer 3 "OSLShader3InputArrayAdapter" (ID 60) :
Shader ColorArrayInputAdapter
 connections in=6 out=1 run_lazily outgoing_connections
  symbols:
param color in0 (used 0 1 read 0 1 write 2147483647 -1) connected
param color in1 (used 2 3 read 2 3 write 2147483647 -1) connected
param color in3 (used 5 6 read 5 6 write 2147483647 -1) connected
oparam color[6] out6 (used 1 9 read 9 9 write 1 8) init [0,9) down-connected
const int $const1 (used 1 1 read 1 1 write 2147483647 -1)
        const: 0
const int $const2 (used 3 3 read 3 3 write 2147483647 -1)
        const: 1
const int $const3 (used 4 4 read 4 4 write 2147483647 -1)
        const: 2
const int $const4 (used 6 6 read 6 6 write 2147483647 -1)
        const: 3
const int $const5 (used 7 7 read 7 7 write 2147483647 -1)
        const: 4
const int $const6 (used 8 8 read 8 8 write 2147483647 -1)
        const: 5
const color $newconst6 (used 4 4 read 4 4 write 2147483647 -1)
        const: 0.75 0.25 0.6
const color $newconst7 (used 7 8 read 7 8 write 2147483647 -1)
        const: 0 1 0.4
  code:
    0: useparam in0     #   BBLOCK-START  (ColorArrayInputAdapter.osl:79)
    1: aassign out6 $const1 (0) in0     #   (ColorArrayInputAdapter.osl:79)
    2: useparam in1     #   (ColorArrayInputAdapter.osl:79)
    3: aassign out6 $const2 (1) in1     #   (ColorArrayInputAdapter.osl:79)
    4: aassign out6 $const3 (2) $newconst6 (0.75 0.25 0.6)      #   CONST  (ColorArrayInputAdapter.osl:79)
    5: useparam in3     #   (ColorArrayInputAdapter.osl:79)
    6: aassign out6 $const4 (3) in3     #   (ColorArrayInputAdapter.osl:79)
    7: aassign out6 $const5 (4) $newconst7 (0 1 0.4)    #   CONST  (ColorArrayInputAdapter.osl:79)
    8: aassign out6 $const6 (5) $newconst7 (0 1 0.4)    #   CONST  (ColorArrayInputAdapter.osl:79)
(main)
    9: useparam out6    #   BBLOCK-START
    10: end     #   CONST
  connections upstream:
    float in0 upconnected from layer 0 (OSLShader)     float globalU
    float in1 upconnected from layer 0 (OSLShader)     float globalU
    color in3 upconnected from layer 2 (OSLShader2)     color c
    float in3 upconnected from layer 2 (OSLShader2)     float c
    float in3 upconnected from layer 2 (OSLShader2)     float c
    float in3 upconnected from layer 2 (OSLShader2)     float c
```

whereas the "Before optimizing layer 3 "OSLShader3InputArrayAdapter" (ID 60)" shows 32 outputs and 500 assignments.

I guess there is some question of noticeable startup overhead if you had an extremely light render with a massive number of spline shaders, but the actual rendering shouldn't be any issue.  And least using the adaptor, we only pay the optimization cost for splines that actually have inputs.